File: BSCM64.rs
// Bounded Smooth Collatz Machine (BSCM) — Engineering Version (Rust Optimized)
// Author: Takeo Yamamoto
// License: Apache 2.0

pub const N: u64 = u64::MAX; // 18446744073709551615

// ─────────────────────────────────────────────────────────────────
// 1. Core transition function δ
// ─────────────────────────────────────────────────────────────────

/// Engineering δ — State-reducing transition.
/// Optimized with bitwise operations to prevent integer overflow.
#[inline]
pub fn bscm_delta(s: u64) -> u64 {
    if s % 2 == 0 {
        s >> 1
    } else {
        // s が u64::MAX のとき、(s + 1) / 2 を直接行うとオーバーフローする。
        // 代数的に同値な 「(s / 2) + 1」 に変形することで、オーバーフローを完全に回避。
        (s >> 1) + 1
    }
}

// ─────────────────────────────────────────────────────────────────
// 2. Control interface
// ─────────────────────────────────────────────────────────────────

/// Single control step.
/// Absorbs arbitrary external input via wrapping addition,
/// then applies δ. Overflow is structurally impossible.
#[inline]
pub fn bscm_control_step(current_state: u64, external_input: u64) -> u64 {
    let s_prime = current_state.wrapping_add(external_input);
    bscm_delta(s_prime)
}

/// Execute a control trace over a sequence of external inputs.
pub fn bscm_control_exec(initial_state: u64, inputs: &[u64]) -> u64 {
    inputs.iter().fold(initial_state, |state, &input| {
        bscm_control_step(state, input)
    })
}

// ─────────────────────────────────────────────────────────────────
// 3. Tests (mirrors Lean theorems)
// ─────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // Theorem 1: state-space invariance (all outputs are valid u64)
    #[test]
    fn test_delta_bounded() {
        for s in [0u64, 1, 2, 3, N, N - 1, N / 2, 12345678901234] {
            // All u64 values are automatically <= u64::MAX, so we just verify no panic
            let result = bscm_delta(s);
            assert!(result <= N);
        }
    }

    // Theorem 2: single-step robust boundedness
    #[test]
    fn test_control_step_bounded() {
        let cases = [(0u64, 0u64), (N, N), (12345u64, u64::MAX), (0u64, u64::MAX)];
        for (state, input) in cases {
            let result = bscm_control_step(state, input);
            assert!(result <= N);
        }
    }

    // Theorem 3: sequence-level safety invariance
    #[test]
    fn test_exec_bounded() {
        let inputs = vec![u64::MAX, u64::MAX, 99999u64, 0u64, u64::MAX];
        let result = bscm_control_exec(u64::MAX, &inputs);
        assert!(result <= N);
    }

    // Reduction property: all branches shrink state (except boundary s = 1)
    #[test]
    fn test_delta_reduces() {
        // s = 1 のときは 1 に収束するため、2以上で検証
        // 奇数では (s >> 1) + 1 = (s - 1) / 2 + 1 = (s + 1) / 2 < s
        // 偶数では s >> 1 < s
        for s in [2u64, 4, 100, 1000, 1000000] {
            assert!(bscm_delta(s) < s, "Failed for s = {}", s);
        }
    }

    // Edge case test for u64::MAX to ensure no overflow panic occurs
    #[test]
    fn test_max_value_boundary() {
        // 元のコードではここでパニックが発生していた
        let res = bscm_delta(u64::MAX);
        // u64::MAX は奇数なので (u64::MAX >> 1) + 1 = 2^63 になるべき
        assert_eq!(res, 1u64 << 63, "Expected 2^63 for u64::MAX");
    }

    // Additional verification: boundary behavior
    #[test]
    fn test_delta_at_one() {
        // s = 1 は収束点（1 のまま）
        assert_eq!(bscm_delta(1), 1);
    }

    // Test control step with overflow-prone inputs
    #[test]
    fn test_control_step_no_panic() {
        // These combinations would cause overflow in naive implementations
        let _ = bscm_control_step(u64::MAX, u64::MAX);
        let _ = bscm_control_step(u64::MAX / 2 + 1, u64::MAX / 2 + 1);
    }
}

File: BSCM64.c
#include <stdint.h>
#include <stdbool.h>

/**
 * @brief BSCMシステムの状態を保持する構造体 (定数空間 O(1))
 */
typedef struct {
    uint64_t current_state;
    bool is_halted;
} BSCM_System;

/**
 * @brief BSCMシステムの初期化
 * @param system 初期化対象のシステム構造体へのポインタ
 * @param initial_state 任意の64ビット初期状態
 */
void bscm_init(BSCM_System *system, uint64_t initial_state) {
    if (system == NULL) return;
    
    // 初期値が0であっても1であっても安全にハンドリング
    system->current_state = (initial_state == 0) ? 1 : initial_state;
    system->is_halted = (system->current_state == 1);
}

/**
 * @brief BSCMコア遷移関数 δ (定数時間 O(1))
 * * どのような破壊的入力(external_input)が飛び込んできても、
 * 絶対にオーバーフローやパニックを起こさず、情報を単調減少（縮退）させます。
 */
void bscm_step(BSCM_System *system, uint64_t external_input) {
    if (system == NULL || system->is_halted) return;

    // 1. 外部入力のカオスビットを安全に結合 (ラップアラウンド加算)
    // どんな巨大な値が来ても、C言語仕様上の未定義動作(UB)を起こさない符号なし安全加算
    uint64_t s = system->current_state + external_input;

    // 2. 状態遷移の分岐 (コラッツ構造の平滑化・外科手術)
    if (s % 2 == 0) {
        // 偶数：1ビット右シフト
        system->current_state = s >> 1;
    } else {
        // 奇数：(s >> 1) + 1 の等価変形により、u64::MAX時でも絶対にオーバーフローしない
        // オリジナルの「3x+1」のような爆発を構造的に絶滅
        system->current_state = (s >> 1) + 1;
    }

    // 3. 終了条件（唯一の不動点 1 への着陸チェック）
    if (system->current_state <= 1) {
        system->current_state = 1;
        system->is_halted = true;
    }
}

/**
 * @brief 外部割り込み・DoSフィルター用ラッパー（最大64ステップで必ず1に収束）
 * * 国家級サイバー攻撃（大量のDoSパケット）を浴びても、
 * この関数全体の最悪実行時間は完全に一定であり、システムを絶対にハングアップさせません。
 */
uint64_t bscm_process_until_halt(uint64_t initial_state, uint64_t external_input) {
    BSCM_System system;
    bscm_init(&system, initial_state);

    // 理論上、64ビット空間では最大64回（線形時間）のループで1に収束する
    // 万が一のハードウェアビット反転に備え、静的な上限付きループ（無限ループの物理的排除）
    for (int i = 0; i < 64; i++) {
        if (system.is_halted) {
            break;
        }
        bscm_step(&system, external_input);
        
        // 1ステップ処理した後は、次のステップでは外部入力を0として
        // 純粋な縮退（収束）フェーズへ移行させる（ノイズの減衰構造）
        external_input = 0;
    }

    return system.current_state; // どのような初期値からも、必ず「1」が返る
}
